### PR TITLE
Use classes as namespaces

### PR DESCRIPTION
**Pull request by mlutze**

### CODE REVIEW

File: main/src/ca/uwaterloo/flix/language/ast/NamedAst.scala [hunk 1]
```diff
@@ -18,15 +18,18 @@ package ca.uwaterloo.flix.language.ast
...
 object NamedAst {
 
+  sealed trait DefOrSig {
+    val loc: SourceLocation
+  }
+
   case class Root(classes: Map[Name.NName, Map[String, NamedAst.Class]],
                   instances: Map[Name.NName, Map[String, List[NamedAst.Instance]]],
-                  defs: Map[Name.NName, Map[String, NamedAst.Def]],
+                  defsAndSigs: Map[Name.NName, Map[String, NamedAst.DefOrSig]],
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/NamedAst.scala:32]: Maybe we need a common name for a Def or Sig?

Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/ast/NamedAst.scala:32]: Yes, that would be nice. Any ideas?

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/ast/NamedAst.scala:32]: Not really. `Member?` but I dont like it.

File: main/src/ca/uwaterloo/flix/language/ast/NamedAst.scala [hunk 2]
```diff
@@ -18,15 +18,18 @@ package ca.uwaterloo.flix.language.ast
 
 import ca.uwaterloo.flix.language.ast
 import ca.uwaterloo.flix.language.ast.Ast.{Denotation, Source}
-import ca.uwaterloo.flix.util.collection.MultiMap
 
 import scala.collection.immutable.List
 
 object NamedAst {
 
+  sealed trait DefOrSig {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/NamedAst.scala:26]: move this definition after case class Root but before Def and Sig

File: main/src/ca/uwaterloo/flix/language/ast/NamedAst.scala [hunk 3]
```diff
@@ -38,9 +41,9 @@ object NamedAst {
...
 
-  case class Sig(doc: Ast.Doc, ann: List[NamedAst.Annotation], mod: Ast.Modifiers, sym: Symbol.SigSym, tparams: List[NamedAst.TypeParam], fparams: List[NamedAst.FormalParam], sc: NamedAst.Scheme, eff: NamedAst.Type, loc: SourceLocation)
+  case class Sig(doc: Ast.Doc, ann: List[NamedAst.Annotation], mod: Ast.Modifiers, sym: Symbol.SigSym, tparams: List[NamedAst.TypeParam], fparams: List[NamedAst.FormalParam], sc: NamedAst.Scheme, eff: NamedAst.Type, loc: SourceLocation) extends DefOrSig
 
-  case class Def(doc: Ast.Doc, ann: List[NamedAst.Annotation], mod: Ast.Modifiers, sym: Symbol.DefnSym, tparams: List[NamedAst.TypeParam], fparams: List[NamedAst.FormalParam], exp: NamedAst.Expression, sc: NamedAst.Scheme, eff: NamedAst.Type, loc: SourceLocation)
+  case class Def(doc: Ast.Doc, ann: List[NamedAst.Annotation], mod: Ast.Modifiers, sym: Symbol.DefnSym, tparams: List[NamedAst.TypeParam], fparams: List[NamedAst.FormalParam], exp: NamedAst.Expression, sc: NamedAst.Scheme, eff: NamedAst.Type, loc: SourceLocation) extends DefOrSig
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/NamedAst.scala:46]: I think Def should occur just above sig

File: main/src/ca/uwaterloo/flix/language/ast/Symbol.scala [hunk 4]
```diff
@@ -301,7 +301,7 @@ object Symbol {
   /**
     * Signature Symbol.
     */
-  final class SigSym(val clazz: Symbol.ClassSym, val name: String, val loc: SourceLocation) {
+  final class SigSym(val clazz: Symbol.ClassSym, val namespace: List[String], val name: String, val loc: SourceLocation) {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/Symbol.scala:304]: Is the namespace not derivable from the class sym? What is this used for?

File: main/test/ca/uwaterloo/flix/language/phase/TestNamer.scala [hunk 5]
```diff
@@ -139,7 +127,66 @@ class TestNamer extends FunSuite with TestUtils {
...
+         |class A[a] {
+         |  def f(x: a): Int
+         |}
+         |
+         |namespace A {
+         |  def f(): Int = 21
+         |}
+       """.stripMargin
+    val result = compile(input, DefaultOptions)
+    expectError[NameError.DuplicateDefOrSig](result)
```
Comment (magnus-madsen) [inline main/test/ca/uwaterloo/flix/language/phase/TestNamer.scala:170]: I would like a few more test cases like this; with nested namespaces and classes.

File: main/test/ca/uwaterloo/flix/language/phase/TestNamer.scala [hunk 6]
```diff
@@ -139,7 +127,66 @@ class TestNamer extends FunSuite with TestUtils {
...
+  }
+
+  test("DuplicateDefOrSig.10") {
```
Comment (magnus-madsen) [inline main/test/ca/uwaterloo/flix/language/phase/TestNamer.scala:173]: OK; like this one. But maybe one where there is just one namespace.

File: main/src/ca/uwaterloo/flix/language/phase/Namer.scala [hunk 7]
```diff
@@ -89,12 +88,34 @@ object Namer extends Phase[WeededAst.Program, NamedAst.Root] {
...
+      val defsAndSigs0 = prog0.defsAndSigs.getOrElse(sigNs, Map.empty)
       classes.get(ident.name) match {
         case None =>
           // Case 1: The class does not already exist. Update it.
-          visitClass(decl, uenv0, Map.empty, ns0) map {
-            case clazz@NamedAst.Class(_, _, _, _, sigs0, _) =>
-              prog0.copy(classes = prog0.classes + (ns0 -> (classes + (ident.name -> clazz))))
+          visitClass(decl, uenv0, Map.empty, ns0) flatMap {
+            case clazz@NamedAst.Class(_, _, _, _, sigs, _) =>
+              // check whether each sig is already in the root
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Namer.scala:98]: I don't understand this check. What is meant by "root"?

**Review by magnus-madsen — COMMENTED** [review]
This is quite nice. It simplifies a lot things.

File: main/src/ca/uwaterloo/flix/language/phase/Namer.scala (hunk 7 above)
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/phase/Namer.scala:98]: Leftover from a previous iteration; should be something like "add each signature to the namespace"

**Comment by magnus-madsen** [comment]
Excellent. Always nice to *reduce* the number of lines of code!

**Pull request merged**